File: Unit - 2/Basic Traversal and Search Techniques/C implementation/DFS.c
//Implement DFS algorithm on a graph represented as Adjacency matrix
#include<stdio.h>
#include<stdio.h>
#define MAX 100

int count = 0;
void readMatrix(int matrix[][MAX], int n)
{
    int i, j;
    for(i = 0; i < n; i++)
    {
        for(j = 0; j < n; j++)
        {
            scanf("%d", &matrix[i][j]);
        }
    }
}

void dfs(int v, int graph[][MAX], int visitedNodes[], int nodeCount)
{
    count++;
    visitedNodes[v] = count;
    printf("%d ", v);
    int i;
    for(int w = 0; w < nodeCount; w++)
    {
        if(graph[v][w] != 0 && visitedNodes[w] == 0)
        {  
            dfs(w, graph, visitedNodes, nodeCount);
        }
    }
}

void DFS(int graph[][MAX], int visitedNodes[], int nodeCount)
{
    int i;
    for(i = 0; i < nodeCount; i++)
    {
        if(visitedNodes[i] == 0)
        {
            dfs(i, graph, visitedNodes, nodeCount);
        }
    }
}

void markUnvisited(int visitedNodes[], int nodeCount)
{
    int i;
    for(i = 0; i < nodeCount; i++)
    {
        visitedNodes[i] = 0;
    }
}

int main()
{
    int nodeCount, graph[MAX][MAX], visitedNodes[MAX];
    printf("Enter the number of nodes: ");
    scanf("%d", &nodeCount);
    printf("Enter the adjacency matrix:\n");
    readMatrix(graph, nodeCount);
    markUnvisited(visitedNodes, nodeCount);
    printf("The order in which the nodes are visited is:\n");
    DFS(graph, visitedNodes, nodeCount);
    printf("\n");
    return 0;
}

File: Unit - 2/Basic Traversal and Search Techniques/Java Implementation/DFS.java
//Implement DFS algorithm on a graph represented as Adjacency matrix
import java.util.Scanner;

public class DFS {
    private static int count =  0;

    private static void DFS(int[][] graph, int[] visitedNodes){
        for(int v = 0; v < visitedNodes.length; v++){
            if(visitedNodes[v] == 0){
                dfs(v, graph, visitedNodes);
            }
        }
        System.out.println();
    }

    private static void dfs(int v, int[][] graph, int[] visitedNodes){
        count++;
        visitedNodes[v] = count;
        System.out.print(v + " ");
        for(int w = 0; w < visitedNodes.length; w++){
            if(graph[v][w] == 1 && visitedNodes[w] == 0){
                dfs(w, graph, visitedNodes);
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the number of nodes: ");
        int nodeCount = sc.nextInt();
        int[][] graph = new int[nodeCount][nodeCount];
        int[] visitedNodes = new int[nodeCount];
        System.out.println("Enter the adjacency matrix: ");
        for(int i = 0; i < nodeCount; i++){
            for(int j = 0; j < nodeCount; j++){
                graph[i][j] = sc.nextInt();
            }
        }
        System.out.println("The order in which the nodes are visited is:");
        DFS(graph, visitedNodes);
    }
}